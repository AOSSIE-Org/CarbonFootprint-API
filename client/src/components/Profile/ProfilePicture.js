import React, { Component } from 'react';
import { Card, Icon } from 'semantic-ui-react';
import ProfileEdit from './ProfileEdit'


/* Extended react.Component class as ProfilePicture */

export default class ProfilePicture extends Component {

  /**
   * Constructor for the ProfilePicture class
   * @constructor extends react.Component
   */

  constructor(props){
    super(props);
    this.state = {
      nickname:""
    }
  }

  /** 
   * Enherit function from react.Component to handle after mounting
   *   react component
   */

  componentDidMount(){
    this.props.auth.getProfile((err, profile) => {
      if(!err){
        this.props.auth.getMetaProfile(profile.sub,(err,metaProfile)=>{
          if(!err){
            metaProfile = JSON.parse(metaProfile);
            if(!metaProfile['user_metadata']){
              metaProfile = {};
              metaProfile['user_metadata']={};
            }
            this.setState({nickname:metaProfile["user_metadata"].nickname});
          }
        });
      }
    });
  }

  /** 
   * Enherited function from react.Component to render to DOM object into html
   */

  render() {
     const { auth } = this.props;
    return (
      <Card>
        <div style={styles.imageContainer}>
          <div
            style={{
              width: '40%',
              paddingTop: '40%',
              borderRadius: '50%',
              background: `url(${this.props.profilePicture}) center center / cover`
            }}
          />
        </div>
        <Card.Content style={{fontSize:'15px', wordWrap:'break-word'}}>
          <Card.Header>
            {this.state.nickname || this.props.nickname} <ProfileEdit auth={auth} />
          </Card.Header>
          <Card.Meta>
            {this.props.email}
          </Card.Meta>
        </Card.Content>
      </Card>
    );
  }
}

const styles = {
  imageContainer: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    background: '#3498db',
    height: '110px'
  }
};
