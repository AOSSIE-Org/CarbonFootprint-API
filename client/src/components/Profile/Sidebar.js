import React, { Component } from 'react';
import { Icon, Menu, Card } from 'semantic-ui-react';

/* Extended react.Component class as Sidebar */

export default class Sidebar extends Component {

  /**
   * Constructor for the Sidebar class
   * @constructor extends react.Component
   */

  constructor() {
    super();
    this.state = { activeItem: 'inbox' };
    this.handleItemClick.bind(this);
  }

  /* Function to handle click event on menu item */

  handleItemClick(e, { name }) {
    this.setState({ activeItem: name });
  }

  /** 
   * Enherited function from react.Component to render to DOM object into html
   */

  render() {
    const { activeItem } = this.state;
    return (
      <Menu vertical>
        <Menu.Item
          name="inbox"
          active={activeItem === 'inbox'}
          onClick={this.handleItemClick}
        >
          <span>
            <Icon name="home" /> Account
          </span>
        </Menu.Item>

        {/*<Menu.Item
          name="spam"
          active={activeItem === 'spam'}
          onClick={this.handleItemClick}
        >
          <span>
            <Icon name="user" /> Edit Profile
          </span>
        </Menu.Item>*/}

        <Menu.Item
          name="updates"
          active={activeItem === 'updates'}
          onClick={() => window.location = 'http://docs.carbonhub.xyz'}
        >
          <span>
            <Icon name="newspaper" /> Documentation
          </span>
        </Menu.Item>
      </Menu>
    );
  }
}
