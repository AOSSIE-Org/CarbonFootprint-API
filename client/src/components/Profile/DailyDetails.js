import React, { Component } from "react";
import { Card, Grid, Icon, Button } from "semantic-ui-react";
import InputData from "./InputData";
import { getKey, submitData, getRawData } from "./UtilDatafetch";
import { raw } from "body-parser";

export default class DailyDetails extends Component {
  state = {
    calculation: [],
    apikey: "",
    total: null,
    loading: false,
    rawdata: ""
  };

  addValue = () => {
    this.setState(prev => {
      prev.calculation.push({
        value: null
      });
      return prev;
    });
  };

  componentDidMount() {
    getKey().then(apikey => this.setState({ apikey }));
    getRawData().then(rawdata => this.setState({ rawdata: rawdata.item }));
  }

  changeCalculation = (index, value) => {
    this.setState(prev => {
      prev.calculation[index] = value;
      return prev;
    });
    this.setState(prev => {
      prev.total = parseFloat(
        prev.calculation
          .reduce(
            (accumulator, currentValue) => accumulator + currentValue.emissions,
            0
          )
          .toFixed(2)
      );
      return prev;
    });
  };

  submitToday = () => {
    this.setState({ loading: true });
    submitData(this.state.total, new Date()).then(data => {
      this.setState({ loading: false });
      this.setState({ calculation: [] });
      window.location = "/profile";
      console.log(data);
    });
  };

  render() {
    // console.log(this.state.rawdata);
    const { calculation, apikey, total, loading, rawdata } = this.state;
    return (
      <Grid className="daily-grid">
        <Grid.Row>
          <Grid.Column mobile={16} tablet={9} computer={9}>
            <Card fluid>
              <Card.Content>
                <Card.Header>Your Activity</Card.Header>
              </Card.Content>
              <Card.Content>
                <Grid className="daily-grid-calculate">
                  {calculation.map((i, index) => (
                    <Grid.Row key={index}>
                      <InputData
                        rawdata={rawdata}
                        apikey={apikey}
                        index={index}
                        changeCalculation={this.changeCalculation}
                      />
                    </Grid.Row>
                  ))}
                  <Grid.Row>
                    <Grid.Column mobile={8} computer={8} tablet={8}>
                      <Icon
                        link
                        size="large"
                        name="add"
                        onClick={this.addValue}
                        className="daily-icon"
                      />
                    </Grid.Column>
                    <Grid.Column mobile={8} computer={8} tablet={8}>
                      <Button
                        primary
                        disabled={this.state.total === null}
                        loading={loading}
                        onClick={this.submitToday}
                        floated="right"
                        className="profile-button"
                      >
                        Total {isNaN(total) ? "" : total}
                      </Button>
                    </Grid.Column>
                  </Grid.Row>
                </Grid>
              </Card.Content>
            </Card>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }
}
