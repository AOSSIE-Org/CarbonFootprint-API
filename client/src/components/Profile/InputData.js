import React, { Component } from 'react';
import { Grid, Input, Select, Button, Header, Dropdown, Message } from 'semantic-ui-react';
import { flightData, trainData, vehicleData, poultryData, appliancesData } from './UtilDatafetch';
import './Profile.css';

export default class InputData extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: null,
      params: [],
      data: {},
      loading: false,
      requiredFields: ['origin', 'destination'],
      errorMessage: "",
      errorActive: false
    };
  }

  setValue = (e, { value }) => {
    this.setState({ value, params: [...this.paramSource[value].params] });
  };

  inputChange = (e, comps) => {
    e.persist();

    this.setState(prev => {
      prev.data[e.target.id || comps.id] = e.target.value || e.target.textContent;
      return prev;
    });

    this.props.changeCalculation(this.props.index, {
      data: this.state.data,
      value: this.state.value
    });
  };

  calculate = () => {
    const {
      origin,
      destination,
      type,
      model,
      passengers,
      mileage,
      region,
      quantity,
      appliance,
      running_time
    } = this.state.data;
    this.setState({ loading: true });
    switch (this.state.value) {
      case 0:
        appliancesData(this.props.apikey, appliance, quantity, running_time, region).then(data => {
          if (data.success) {
            this.setState({ emissions: data.emissions, errorMessage: "", loading: false });
            this.props.changeCalculation(this.props.index, {
              emissions: this.state.emissions.CO2
            });
          }
          else {
            this.setState({ emissions: undefined, errorMessage: data.message, loading: false });
          }
        });
        break;
      case 1:
        poultryData(this.props.apikey, type, region, quantity).then(data => {
          if (data.success) {
            this.setState({ emissions: data.emissions, loading: false, errorMessage: "" });
            this.props.changeCalculation(this.props.index, {
              emissions: this.state.emissions.CO2
            });
          }
          else {
            this.setState({ emissions: undefined, errorMessage: data.message, loading: false });
          }
        });
        break;
      case 2:
        flightData(this.props.apikey, origin, destination, type, model, passengers).then(data => {
          if (data.success) {
            this.setState({ emissions: data.emissions, loading: false, errorMessage: "" });
            this.props.changeCalculation(this.props.index, {
              emissions: this.state.emissions.CO2
            });
          }
          else {
            this.setState({ emissions: undefined, errorMessage: data.message, loading: false });
          }
        });
        break;
      case 3:
        vehicleData(this.props.apikey, origin, destination, type, mileage).then(
          data => {
            if (data.success) {
              this.setState({ emissions: data.emissions, errorMessage: "", loading: false });
              this.props.changeCalculation(this.props.index, {
                emissions: this.state.emissions.CO2
              });
            }
            else {
              this.setState({ emissions: undefined, errorMessage: data.message, loading: false });
            }
          });
        break;
      case 4:
        trainData(this.props.apikey, origin, destination, type, passengers).then(data => {
          if (data.success) {
            this.setState({ emissions: data.emissions, loading: false, errorMessage: "" });
            this.props.changeCalculation(this.props.index, {
              emissions: this.state.emissions.CO2
            });
          }
          else {
            this.setState({ emissions: undefined, errorMessage: data.message, loading: false });
          }
        });
        break;
      default:
        this.setState({ loading: false })
        console.log('no option');
        break;
    }
  };

  render() {
    // console.log(this.state.errorMessage);
    const { params, loading } = this.state;
    return (
      <Grid className="input-grid">
        <Grid.Row>
          <Grid.Column mobile={8} computer={7} tablet={7}>
            <Select
              fluid
              placeholder="Activity..."
              search
              selection
              onChange={this.setValue}
              options={this.options}
            />
          </Grid.Column>
          {this.state.emissions !== undefined && (
            <Grid.Column
              className="counter column-emission-value"
              floated="right"
              mobile={4}
              computer={4}
              tablet={4}>
              <Header as="h3" className="input-header">
                {this.state.emissions.CO2.toFixed(2)} kg
              </Header>
            </Grid.Column>
          )}
        </Grid.Row>
        <Grid.Row>
          <Grid doubling columns={3}>
            {params.map((comps, index) => (
              <Grid.Column key={index}>
                {(this.paramSource[this.state.value][comps]) ? (
                  <Dropdown
                    defaultOpen={true}
                    fluid
                    selection
                    placeholder={`${comps}`}
                    id={comps}
                    onChange={(e, comps) => { this.inputChange(e, comps) }}
                    options={this.paramSource[this.state.value][comps]}
                    style={{ marginTop: '-22px' }}
                  />
                )
                  : (<Input
                    error={((this.state.requiredFields.includes(comps)) && !this.state.data[comps] && this.state.errorActive) ? true : false}
                    fluid
                    placeholder={`${comps}`}
                    id={comps}
                    onChange={this.inputChange}
                    style={{ marginTop: '-22px' }}
                  />
                  )}
              </Grid.Column>
            ))}
            {params.length !== 0 && (
              <Grid.Column className="column-calculate">
                <Button
                  className="profile-button"
                  fluid
                  disabled={loading}
                  loading={loading}
                  onClick={this.calculate}>
                  Calculate
                </Button>
              </Grid.Column>
            )}
          </Grid>
        </Grid.Row>
        {(this.state.errorMessage) ? (<Message color='red' >{this.state.errorMessage}</Message>) : null}
      </Grid>
    );
  }
  paramSource = [
    {
      title: 'appliances',
      params: ['appliance', 'region', 'quantity', 'runnning_time'],
      appliance: this.props.rawdata.applianceTypes.map((i, index) => ({
        key: index,
        value: index,
        text: i
      })),
    },
    {
      title: 'poultry',
      params: ['type', 'region', 'quantity'],
      type: this.props.rawdata.poultryTypes.map((i, index) => ({
        key: index,
        value: index,
        text: i
      })),
    },
    {
      title: 'flight',
      params: ['origin', 'destination', 'type', 'model', 'passengers'],
      model: this.props.rawdata.flightTypes.map((i, index) => ({
        key: index,
        value: index,
        text: i
      })),
    },
    {
      title: 'vehicle',
      params: ['type', 'origin', 'destination', 'mileage'],
      type: this.props.rawdata.vehicleTypes.map((i, index) => ({
        key: index,
        value: index,
        text: i
      })),
    },
    {
      title: 'trains',
      params: ['type', 'origin', 'destination', 'passengers'],
      type: this.props.rawdata.trainTypes.map((i, index) => ({
        key: index,
        value: index,
        text: i
      })),
    }
  ];

  options = this.paramSource.map((i, index) => ({
    key: i.title,
    value: index,
    text: i.title.charAt(0).toUpperCase() + i.title.slice(1)
  }));
}

