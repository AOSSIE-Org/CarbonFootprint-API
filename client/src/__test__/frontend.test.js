import React from 'react'
import ReactDOM from 'react-dom'
import { BrowserRouter as Router } from 'react-router-dom'
import { render, cleanup, fireEvent } from '@testing-library/react'
import ProfilePicture from '../components/Profile/ProfilePicture'
import Sidebar from '../components/Profile/Sidebar'
import InputData from '../components/Profile/InputData'
import DailyDetails from '../components/Profile/DailyDetails';

afterEach(cleanup)

it('renders ProfilePicture component without crashing', () => {
	const div = document.createElement('div');
	ReactDOM.render(<ProfilePicture />, div)
})

it('compare snapshot of ProfilePicture component', () => {
	const { asFragment } = render(<ProfilePicture />)
	expect(asFragment(<ProfilePicture />)).toMatchSnapshot()
})

it('renders Sidebar component without crashing', () => {
	const div = document.createElement('div');
	ReactDOM.render(<Router><Sidebar /></Router>, div)
})

it('compare snapshot of Sidebar component', () => {
	const { asFragment } = render(<Router><Sidebar />)</Router>)
	expect(asFragment(<Router><Sidebar /></Router>)).toMatchSnapshot()
})

it('renders DailyDetails component without crashing', () => {
	const div = document.createElement('div');
	ReactDOM.render(<DailyDetails />, div)
})

it('renders InputData component without crashing', () => {
	const div = document.createElement('div');
	ReactDOM.render(<InputData />, div)
})


// it('checking functionality of DailyDetails Component', () => {
// 	const { getByTestId } = render(<DailyDetails />)
// 	fireEvent.click(getByTestId('plus-button'))
// 	expect(<InputData />).toBeInTheDocument
// })




