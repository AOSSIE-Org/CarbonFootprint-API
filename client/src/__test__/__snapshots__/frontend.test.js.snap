// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compare snapshot of ProfilePicture component 1`] = `
<DocumentFragment>
  <div
    class="ui fluid raised card"
  >
    <div
      style="display: flex; justify-content: center; align-items: center; width: 100%; background: rgb(52, 152, 219); height: 110px;"
    >
      <div
        style="width: 40%; padding-top: 40%; border-radius: 50%;"
      />
    </div>
    <div
      class="content"
      style="font-size: 15px; word-wrap: break-word;"
    >
      <div
        class="header"
      >
         
        <div
          style="display: inline;"
        >
          <i
            aria-hidden="true"
            class="write icon"
            style="float: right; font-size: 15px; cursor: pointer;"
          />
        </div>
      </div>
      <div
        class="meta"
        style="display: inline-flex; overflow-x: hidden; width: 100%;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`compare snapshot of Sidebar component 1`] = `
<DocumentFragment>
  <div
    class="ui fluid vertical menu"
    style="margin-bottom: 15px;"
  >
    <a
      class="link item"
      href="/profile"
    >
      <span
        style="color: rgb(65, 131, 196);"
      >
        <i
          aria-hidden="true"
          class="home icon"
        />
         Account
      </span>
    </a>
    <a
      class="link item"
      href="/TransportComparer"
    >
      <span
        style="color: rgb(65, 131, 196);"
      >
        <i
          aria-hidden="true"
          class="train icon"
        />
         Transport Comparer
      </span>
    </a>
    <div
      class="item"
    >
      <a
        href="http://docs.carbonhub.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span>
          <i
            aria-hidden="true"
            class="newspaper icon"
          />
           API Documentation
        </span>
      </a>
    </div>
    <div
      class="item"
    >
      <a
        href="/DataUpload"
      >
        <span>
          <i
            aria-hidden="true"
            class="upload icon"
          />
           Data Upload
        </span>
      </a>
    </div>
    <div
      class="item"
    >
      <a
        href="http://localhost:3080/api/docs"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span>
          <i
            aria-hidden="true"
            class="wpforms icon"
          />
           Swagger Documentation
        </span>
      </a>
    </div>
    <div
      class="item"
    >
      <a
        href="/dataVerify"
      >
        <span>
          <i
            aria-hidden="true"
            class="check circle icon"
          />
           Verify Data
        </span>
      </a>
    </div>
  </div>
  )
</DocumentFragment>
`;
